fix(cli): ensure clean exit on provider/model errors

imaurer opened the pull request:
Summary
- Add actionable error formatting for ProviderModelNotFoundError and ProviderInitError.
- Ensure the CLI exits cleanly by confirming teardown of process-scoped state (already present on `dev`).

Context
- The hang was caused by process-scoped services (watchers, etc.) not being disposed when errors occurred during `opencode run`.
- The teardown via try/finally in bootstrap is already on `dev`; this PR contributes the friendly error messages to guide users when provider/model resolution fails.

Reproduction
```
opencode run --model typo/claude-haiku-4-5 "hi there"
```
Observed (pre-fix): noisy stack + process required Ctrl+C.

Change
- packages/opencode/src/cli/error.ts
  - Format provider/model errors into actionable guidance
- packages/opencode/src/cli/bootstrap.ts
  - No diff vs `dev` (teardown exists upstream)

Notes
- A prior change (#3083) sought to avoid hangs via explicit exit. The upstream `dev` now guarantees teardown via try/finally; this PR complements that by improving the UX for a common misconfiguration.

Testing
- With upstream teardown + this change, the above repro exits with code 1 and shows actionable guidance.

imaurer committed:
fix(cli): always dispose instance on error to prevent hanging; add friendly ProviderModelNotFoundError/InitError messages\n\n- Wrap bootstrap callback in try/finally to guarantee Instance.dispose()\n- Format provider/model errors into actionable guidance (opencode models, config)\n\nRepro: running `opencode run --model typo/claude-haiku-4-5` prints stack and hangs until SIGINT due to lingering watchers.\nFix: disposing Instance tears down watchers/subscriptions, allowing process to exit.\n\nNotes: Prior attempt (#3083) explicitly exited; this approach addresses root cause without forcing exit and improves UX for common misconfigurations.

imaurer committed:
chore(cli): resolve merge conflict with dev in bootstrap; keep try/finally with explicit return

imaurer committed:
Merge branch 'dev' into fix/cli-clean-exit-on-model-errors

rekram1-node commented on packages/opencode/src/cli/error.ts:
```diff
@@ -1,11 +1,23 @@
 
 export function FormatError(input: unknown) {
   if (MCP.Failed.isInstance(input))
     return `MCP server "${input.data.name}" failed. Note, opencode does not support MCP authentication yet.`
+  if (Provider.ModelNotFoundError.isInstance(input)) {
+    const { providerID, modelID } = input.data
+    return [
+      `Model not found: ${providerID}/${modelID}`,
+      `Try: \`opencode models\` to list available models`,
+      `Or check your config (opencode.json) provider/model names`,
+    ].join("\n")
+  }
+  if (Provider.InitError.isInstance(input)) {
+    return `Failed to initialize provider "${input.data.providerID}". Check credentials and configuration.`
```
do u think we can try to do `did u mean X` too? I bet we can try to find best match

imaurer committed:
feat(cli): suggest closest provider/model on not found ("Did you mean…")\n\nSummary\n- Add fuzzy suggestions to ProviderModelNotFoundError with up to 3 candidates\n- Normalize punctuation (e.g., 4.5 vs 4-5) and case to better match common typos\n- Support model-only input (no provider) by searching across all providers\n- Enhance CLI error formatter to display suggestions when present\n\nImplementation\n- provider.ts: use fuzzysort; add normalization by stripping non-alphanumerics; search by key for robust matches\n- provider.ts: when provider is unknown and model is empty, treat token as unqualified model and search across all providers' models; otherwise suggest provider matches\n- error.ts: print "Did you mean: <provider/model>, …" when suggestions exist\n\nExamples\n1) Typo in model ID\n   $ bun run ./src/index.ts run --model anthropic/claude-haiu-4-5 "hi"\n   Error: Model not found: anthropic/claude-haiu-4-5\n   Did you mean: anthropic/claude-haiku-4-5, anthropic/claude-haiku-4-5-20251001\n   Try: zai-coding-plan/glm-4.5-flash
zai-coding-plan/glm-4.5
zai-coding-plan/glm-4.5-air
zai-coding-plan/glm-4.5v
zai-coding-plan/glm-4.6
opencode/big-pickle
opencode/grok-code
anthropic/claude-opus-4-0
anthropic/claude-3-5-sonnet-20241022
anthropic/claude-opus-4-1
anthropic/claude-haiku-4-5
anthropic/claude-3-5-sonnet-20240620
anthropic/claude-3-5-haiku-latest
anthropic/claude-3-opus-20240229
anthropic/claude-sonnet-4-5
anthropic/claude-sonnet-4-5-20250929
anthropic/claude-sonnet-4-20250514
anthropic/claude-opus-4-20250514
anthropic/claude-3-5-haiku-20241022
anthropic/claude-3-haiku-20240307
anthropic/claude-3-7-sonnet-20250219
anthropic/claude-3-7-sonnet-latest
anthropic/claude-sonnet-4-0
anthropic/claude-opus-4-1-20250805
anthropic/claude-3-sonnet-20240229
anthropic/claude-haiku-4-5-20251001
openai/gpt-4.1-nano
openai/text-embedding-3-small
openai/gpt-4
openai/o1-pro
openai/gpt-4o-2024-05-13
openai/gpt-4o-2024-08-06
openai/gpt-4.1-mini
openai/o3-deep-research
openai/gpt-3.5-turbo
openai/text-embedding-3-large
openai/gpt-4-turbo
openai/o1-preview
openai/o3-mini
openai/codex-mini-latest
openai/gpt-5-nano
openai/gpt-5-codex
openai/gpt-4o
openai/gpt-4.1
openai/o4-mini
openai/o1
openai/gpt-5-mini
openai/o1-mini
openai/text-embedding-ada-002
openai/o3-pro
openai/gpt-4o-2024-11-20
openai/o3
openai/o4-mini-deep-research
openai/gpt-4o-mini
openai/gpt-5
openai/gpt-5-pro to list available models\n   Or check your config (opencode.json) provider/model names\n\n2) Dot vs dash (punctuation normalization)\n   $ bun run ./src/index.ts run --model anthropic/claude-haiku-4.5 "hi"\n   Error: Model not found: anthropic/claude-haiku-4.5\n   Did you mean: anthropic/claude-haiku-4-5, anthropic/claude-haiku-4-5-20251001\n   Try: zai-coding-plan/glm-4.5-flash
zai-coding-plan/glm-4.5
zai-coding-plan/glm-4.5-air
zai-coding-plan/glm-4.5v
zai-coding-plan/glm-4.6
opencode/big-pickle
opencode/grok-code
anthropic/claude-opus-4-0
anthropic/claude-3-5-sonnet-20241022
anthropic/claude-opus-4-1
anthropic/claude-haiku-4-5
anthropic/claude-3-5-sonnet-20240620
anthropic/claude-3-5-haiku-latest
anthropic/claude-3-opus-20240229
anthropic/claude-sonnet-4-5
anthropic/claude-sonnet-4-5-20250929
anthropic/claude-sonnet-4-20250514
anthropic/claude-opus-4-20250514
anthropic/claude-3-5-haiku-20241022
anthropic/claude-3-haiku-20240307
anthropic/claude-3-7-sonnet-20250219
anthropic/claude-3-7-sonnet-latest
anthropic/claude-sonnet-4-0
anthropic/claude-opus-4-1-20250805
anthropic/claude-3-sonnet-20240229
anthropic/claude-haiku-4-5-20251001
openai/gpt-4.1-nano
openai/text-embedding-3-small
openai/gpt-4
openai/o1-pro
openai/gpt-4o-2024-05-13
openai/gpt-4o-2024-08-06
openai/gpt-4.1-mini
openai/o3-deep-research
openai/gpt-3.5-turbo
openai/text-embedding-3-large
openai/gpt-4-turbo
openai/o1-preview
openai/o3-mini
openai/codex-mini-latest
openai/gpt-5-nano
openai/gpt-5-codex
openai/gpt-4o
openai/gpt-4.1
openai/o4-mini
openai/o1
openai/gpt-5-mini
openai/o1-mini
openai/text-embedding-ada-002
openai/o3-pro
openai/gpt-4o-2024-11-20
openai/o3
openai/o4-mini-deep-research
openai/gpt-4o-mini
openai/gpt-5
openai/gpt-5-pro to list available models\n   Or check your config (opencode.json) provider/model names\n\n3) Missing provider (model-only input)\n   $ bun run ./src/index.ts run --model big-pickle "hi"\n   Error: Model not found: big-pickle/\n   Did you mean: opencode/big-pickle\n\n4) Correct model after suggestion\n   $ bun run ./src/index.ts run --model opencode/big-pickle "hi"\n   Hi! How can I help you with your opencode project today?\n\nNotes\n- Suggestions are hints only; behavior is unchanged (no auto-selection).\n- This runs locally as part of the CLI error path; performance impact is negligible (small in-memory scans).

actions-user committed:
chore: format code

imaurer committed:
fix(cli): robust suggestions for unknown provider and model-only input; ignore bun build artifacts

actions-user committed:
chore: format code

imaurer committed:
Merge branch 'dev' into fix/cli-clean-exit-on-model-errors

imaurer committed:
Update .gitignore to remove '*.bun-build'

Remove '*.bun-build' from the .gitignore file

imaurer commented:
@rekram1-node

Good idea on the provider/model suggestion capability.

I worked with codex to get this working.

But, honestly, I am not a TS developer and also don't know opencode code base very well.

My goal was to get something that worked for these scenarios:
- typo
- dot not dash
- missing provider

Examples:

```
❯ ./dist/opencode-darwin-arm64/bin/opencode run --model anthropic/claude-haiu-4-5 "hi"

Error: Model not found: anthropic/claude-haiu-4-5
Did you mean: anthropic/claude-haiku-4-5, anthropic/claude-haiku-4-5-20251001
Try: `opencode models` to list available models
Or check your config (opencode.json) provider/model names


❯ ./dist/opencode-darwin-arm64/bin/opencode run --model anthropic/claude-haiku-4.5 "hi"

Error: Model not found: anthropic/claude-haiku-4.5
Did you mean: anthropic/claude-haiku-4-5, anthropic/claude-haiku-4-5-20251001
Try: `opencode models` to list available models
Or check your config (opencode.json) provider/model names


❯ ./dist/opencode-darwin-arm64/bin/opencode run --model big-pickle "hi"

Error: Model not found: big-pickle/
Did you mean: opencode/big-pickle
Try: `opencode models` to list available models
Or check your config (opencode.json) provider/model names


❯ ./dist/opencode-darwin-arm64/bin/opencode run --model anthropic/big-pickle "hi"

Error: Model not found: anthropic/big-pickle
Try: `opencode models` to list available models
Or check your config (opencode.json) provider/model names
```

Hope this is helpful and meets the project's quality standards.
If not, I fully understand and apologize for the PR.

rekram1-node commented:
This is good, I will prolly touch up on it today but we will merge this, thank u

imaurer commented:
Anything else I can do @rekram1-node ?

rekram1-node commented:
Nah all good just bouncing between a few prs, Ill try to merge this one today

rekram1-node committed:
merge dev

rekram1-node committed:
tweak: reduce code

imaurer commented:
Thanks!,
